### PR TITLE
[FEATURE] Add Automatic Mixed Precision option

### PR DESCRIPTION
**Pull request by mwulfman**
## Proposal 
I'm proposing adding an option to enable Automatic Mixed Precision (AMP) in order to improve training throughput with minimal code changes.

## Results
To validate the implementation, I benchmarked training before and after enabling AMP with:

```sh
python mimickit/run.py \
  --mode train \
  --num_envs 2048 \
  --engine_config data/engines/isaac_lab_engine.yaml \
  --env_config data/envs/deepmimic_humanoid_env.yaml \
  --agent_config data/agents/deepmimic_humanoid_ppo_agent.yaml \
  --visualize false
 ```
 
 The only configuration change was enabling in the current branch:
```yaml
     use_mixed_precision: True
```

Using the additional logging introduced in https://github.com/xbpeng/MimicKit/pull/107, I obtained the following results on an NVIDIA GeForce RTX 5070 Ti Laptop GPU.

 
Before Automatic Mixed Precision:
```sh
|   Samples_Per_Second |        3.89e+04 |
```

After Automatic Mixed Precision:
```sh
|   Samples_Per_Second |        6.86e+04 |
```

This corresponds to +70% increase in training throughput.

## Changes done:

- [x] Update `MPOptimizer` to support Automatic Mixed Precision (AMP) with both `"fp16"` and `"bf16"` modes
- [x] Update the agent `_update_actor` and `_update_critic` methods to pass a `loss_factory` to `optimizer.step`, allowing the forward pass and loss computation to run under AMP autocast
- [x] Update the configs while keeping the default behavior unchanged

### CODE REVIEW

**Comment by xbpeng** [comment]
Thanks for submitting this PR, that looks like a really nice speedup! I wonder if it is possible to simplify the control flow. Instead of using a loss factory and calling autocast in the optimizer, we could put autocast around `_update_model` instead:
https://github.com/xbpeng/MimicKit/blob/aeb7de922f02302f709eb2d91705929f1e84e504/mimickit/learning/base_agent.py#L238
That way, all of the operations for model updates will be performed in mixed precision. This could help to simplify the code, and avoid the need for loss factories.

Also, "amp" conflicts with adversarial motion priors. So instead of using `amp` maybe we should go with `mixed_precision`.

I haven't used mixed precision before. But it seems like `GradScaler` is only used if `amp_type == torch.float16`. Is it not necessary for `bf16`?

**Comment by mwulfman** [comment]
Yes, I agree it would be nicer to simplify the code and avoid the `loss_factory` indirection if possible.

My only concern with wrapping the whole `_update_model()` call in autocast is that PyTorch generally recommends using autocast only around the forward pass and loss computation, and leaving the backward / optimizer step outside the autocast context.

One possible alternative would be to move autocast into `_update_actor()` / `_update_critic()` directly, for example:

```python
def _update_critic(self, batch_size, steps):
    info = dict()

    for _ in range(steps):
        batch = self._exp_buffer.sample(batch_size)
        with torch.amp.autocast(
            device_type="cuda",
            enabled=True,
            dtype=torch.bfloat16,
        ):
            loss_info = self._compute_critic_loss(batch)
            loss = loss_info["critic_loss"]

        self._critic_optimizer.step(loss)
        torch_util.add_torch_dict(loss_info, info)

    torch_util.scale_torch_dict(1.0 / steps, info)
    return info
```

The limitation is that this works cleanly for `bfloat16`, but not as well for `float16`, since at this level we do not have a good place to handle `GradScaler` scale / unscale logic. 
For bf16 loss scaling is not needed because, compared to fp16, bf16 has a similar dynamic range to fp32. The downside is that some older/pre-Ampere GPUs do not have efficient/native bf16 support, while they may still support fp16 mixed precision.

So I think the tradeoff is:

1. keep the current `loss_factory` approach if we want to support both fp16 and bf16 properly;
2. simplify the code if we are okay with supporting only bf16 mixed precision.

I am fine with either direction.

**Comment by xbpeng** [comment]
is there a strong reason to support float16? Would bfloat16 be sufficient for most use cases?

**Comment by mwulfman** [comment]
Yes, bfloat16 would be sufficient for most use cases. To my understanding, it is the recommended type for mixed precision.
Supporting float16 would allow users with V100 GPUs to use mixed precision. However, going forward, all GPUs support bfloat16.
I'm happy to simplify the code and support only bfloat16.

**Comment by mwulfman** [comment]
I've updated the proposal to support only bfloat16 for mixed precision. Let me know what you think about it.

**Comment by xbpeng** [comment]
Thanks for the update! that looks much cleaner. Can you also add the mixed precision scope for other loss functions like in AMP and ASE?
https://github.com/xbpeng/MimicKit/blob/adbee3d1bec6bf901348f9f99646cd27000f0d9d/mimickit/learning/amp_agent.py#L153
https://github.com/xbpeng/MimicKit/blob/adbee3d1bec6bf901348f9f99646cd27000f0d9d/mimickit/learning/ase_agent.py#L258
I'll also do some testing on my side and it should be good to go!

**Comment by mwulfman** [comment]
Done, let me know if I missed another one.

**Comment by xbpeng** [comment]
Looks great, thanks for putting this together! Merged!
